### PR TITLE
Repository Listing page driven by Haystack.

### LINKED ISSUE #144: Search: Make the repository listing page powered by search
See #120 

When you visit the home page of a repository  https://goo.gl/TKDuK2 , the results come from the database. 
As soon as you click a facet or search for something  https://goo.gl/5ZVEp5, the datasource become the search index db.

This issue is about consolidating them so that the home page of the repository (the current "listing page") is powered by search index.

Technically, this is described here 

http://stackoverflow.com/questions/10978695/django-haystack-show-results-without-needing-a-search-query

Especially, in one of the answers:

http://stackoverflow.com/a/24200522

### PR DESCRIPTION
**Pull request by ShawnMilo**
Also moves search to its own app.

closes #144

### CODE REVIEW

File: importer/api/__init__.py [hunk 1]
```diff
@@ -10,9 +10,9 @@
 from os.path import join, exists
 from os import remove, listdir
 
-from xbundle import XBundle, DESCRIPTOR_TAGS
-from lxml import etree
 from archive import extract, ArchiveException
+from lxml import etree
+from xbundle import XBundle, DESCRIPTOR_TAGS
```
Comment (ShawnMilo) [inline importer/api/__init__.py:16]: These changes are from alphabetizing the imports.

File: learningresources/api.py [hunk 2]
```diff
@@ -100,12 +98,9 @@ def type_id_by_name(name):
     Returns:
         type_id (int): Primary key of learningresources.LearningResourceType
     """
-    if name in TYPE_LOOKUP:
-        return TYPE_LOOKUP[name]
     with transaction.atomic():
         obj, _ = LearningResourceType.objects.get_or_create(name=name.lower())
-    TYPE_LOOKUP[name] = obj.id
-    return obj.id
+        return obj.id
```
Comment (ShawnMilo) [inline learningresources/api.py:105]: The "caching" was moved because it was breaking tests.

File: lore/settings.py [hunk 3]
```diff
@@ -280,3 +280,4 @@ def get_var(name, default):
         'INDEX_NAME': 'haystack'
     }
 }
+HAYSTACK_SIGNAL_PROCESSOR = 'haystack.signals.RealtimeSignalProcessor'
```
Comment (ShawnMilo) [inline lore/settings.py:284]: This causes the index to be updated whenever a record is saved or deleted. This closes #142, but was unavoidable because otherwise no indexing happened during tests. Anyway, it's only one line.

File: ui/templates/repositories.html [hunk 4]
```diff
@@ -3,13 +3,17 @@
 {% block content %}
```
Comment (ShawnMilo) [inline ui/templates/repositories.html:3]: This file was renamed to match all the URLs.

File: ui/templates/search/indexes/learningresources/learningresource_text.txt [hunk 5]
```diff
@@ -1,3 +1,6 @@
 {{ object.title }}
 {{ object.description }}
 {{ object.LearningResourceType.name }}
+{{ object.Course.course_number }}
+{{ object.Course.Repository.name }}
+{{ object.Course.Repository.id }}
```
Comment (ShawnMilo) [inline ui/templates/search/indexes/learningresources/learningresource_text.txt:6]: The values in this template are what Haystack uses to build index metadata.

File: ui/views.py [hunk 6]
```diff
@@ -3,8 +3,9 @@
 """
 from __future__ import unicode_literals
 
+import logging
+
 from django.contrib.auth.decorators import login_required
-from django.core.paginator import Paginator
```
Comment (ShawnMilo) [inline ui/views.py:7]: Haystack uses Paginator internally.

File: ui/views.py [hunk 7]
```diff
@@ -190,34 +194,31 @@ def create_repo(request):
...
-        "tags": get_user_tags(repo.id),
-        "resources": Paginator(
-            get_resources(repo.id), 20).page(page)
-    }
-    return render(
-        request,
-        "listing.html",
-        context,
-    )
+class RepositoryView(SearchView):
```
Comment (ShawnMilo) [inline ui/views.py:207]: This view, a subclass of a Haystack class-based view, is the meat of this PR. Mostly, it encapsulates search and pagination. Most of the code here is to heal with the slug for user permissions.

**Comment by ShawnMilo** [comment]
I don't know why this failed in Travis. It passes locally. It had a connection failure to Elasticsearch, but it passed in the tests before. I'm hoping it's just a fluke wit Travis's servers. I'll make a small change and re-push to see if it works this time.

**Comment by giocalitri** [comment]
The tests are failing in Travis (and locally) because elasticsearch is not available.
On Travis we should enable elasticsearch: @carsongee is [this](http://docs.travis-ci.com/user/database-setup/#ElasticSearch) the right documentation?

**Comment by giocalitri** [comment]
After this is solved, let's :shipit:

File: ui/forms.py [hunk 8]
```diff
@@ -36,14 +36,10 @@ def __init__(self, *args, **kwargs):
     def clean_course_file(self):
         """Only certain extensions are allowed."""
         upload = self.cleaned_data["course_file"]
-        log.debug("checking filename %s", upload.name)
         for ext in (".zip", ".tar.gz", ".tgz"):
```
Comment (giocalitri) [inline ui/forms.py:40]: Should we move this list of extensions in a more visible positions?

File: importer/api/__init__.py [hunk 9]
```diff
@@ -43,7 +43,7 @@ def import_course_from_file(filename, repo_id, user_id):
     try:
         extract(path=filename, to_path=tempdir, method="safe")
     except ArchiveException as ex:
-        log.debug("failed to extract: %s", ex)
+        log.error("failed to extract: %s", ex)
```
Comment (carsongee) [inline importer/api/__init__.py]: This seems out of scope for this PR, also, if you use log.execption it prints the stack trace and you don't have to pass the exception into it.

File: importer/api/__init__.py [hunk 10]
```diff
@@ -74,7 +74,7 @@ def import_course_from_path(path, repo_id, user_id):
     """
     bundle = XBundle()
     bundle.import_from_directory(path)
-    return import_course(bundle, repo_id, user_id)
+    import_course(bundle, repo_id, user_id)
```
Comment (carsongee) [inline importer/api/__init__.py]: again, this seems strange to have in a haystack PR

File: importer/tests/test_views.py [hunk 11]
```diff
@@ -66,5 +66,4 @@ def test_invalid_form(self):
         )
         self.assertTrue(resp.status_code == HTTP_OK)
         body = resp.content.decode("utf-8")
-        log.debug(body)
```
Comment (carsongee) [inline importer/tests/test_views.py:161]: nothing wrong with leaving this in.  They only show when that test fails and they are handy to have

File: importer/tests/test_forms.py [hunk 12]
```diff
@@ -50,18 +50,14 @@ def test_extensions(self):
...
     def attempt_extension(self, ext):
         """Try to upload a file with a given extension."""
         upload = get_upload_file(ext)
-        log.debug("attempt_extension %s, %s", ext, upload.name)
         form = UploadForm(
             {"repository": self.repo.id},
             {"course_file": upload},
         )
         is_valid = form.is_valid()
-        log.debug("%s was valid: %s", upload.name, is_valid)
```
Comment (carsongee) [inline importer/tests/test_forms.py:66]: same here on leaving these in and scope.  Feel free to submit these as a sep PR though

File: importer/api/__init__.py (hunk 10 above)
Comment (ShawnMilo) [inline importer/api/__init__.py]: It was just cleaning as I went along. The called function doesn't return a value, so it was just misleading. Is there any reason _not_ to change it here?

File: importer/tests/test_forms.py (hunk 12 above)
Comment (ShawnMilo) [inline importer/tests/test_forms.py:66]: I'd prefer to just leave this. Is there any reason that would be bad?

File: importer/tests/test_views.py (hunk 11 above)
Comment (ShawnMilo) [inline importer/tests/test_views.py:161]: Right, since it won't pollute output the rest of the time, I'll restore that.

File: importer/api/__init__.py (hunk 10 above)
Comment (carsongee) [inline importer/api/__init__.py]: Just for a clean commit history.  i.e. git blame/history will show this was part of haystack, which could be confusing.  Generally the one commit, one PR, one logical change methodology.  So several small one or two line PRs are easy to review and merge and the history looks good, but larger lumbering ones get muddy.

File: ui/forms.py (hunk 8 above)
Comment (ShawnMilo) [inline ui/forms.py:40]: I think the form validation is the best place for it, it's not used anywhere else, and that would require another import.

However, it should probably be in the UI on the import page user-facing documentation. That would justify having it in a setting somewhere else. Based on @carsongee's comments for other parts of this PR that would be out of scope, so please make a separate issue for that.

File: importer/api/__init__.py (hunk 10 above)
Comment (ShawnMilo) [inline importer/api/__init__.py]: Okay, that makes sense. It would be trivial for me to create and push those tiny PRs while I'm working, and I'm sure they'd get pulled into `master` the same day, so same end result and cleaner.

File: ui/forms.py (hunk 8 above)
Comment (carsongee) [inline ui/forms.py:40]: Just to add to that as an issue, I'd also like to add libmagic checks for the file type.  Those two would go well together when you write the issue

Comment (giocalitri) [inline ui/forms.py:40]: I'll open an issue

**Comment by ShawnMilo** [comment]
Note: The only thing not addressed in this branch is moving the config out of `lore/settings.py` as Carson suggested during the stand-up. Right now it's not working, and PR #154 is up to investigate. If it can't be moved for some reason, then this is done if it passes Travis.

**Comment by ShawnMilo** [comment]
Yay, thanks @giocalitri -- tests are passing in Travis now thanks to your help.

**Comment by giocalitri** [comment]
will do one last pass and merge

**Comment by ShawnMilo** [comment]
@giocalitri: Sounds great, but please check with @carsongee about #154 just in case.

**Comment by giocalitri** [comment]
LGTM: @carsongee if #154 is fine, feel free to :shipit: (or I can press the button if you prefer)

**Comment by carsongee** [comment]
You can merge this now, since #154 is open
